### PR TITLE
Debug CI

### PR DESCRIPTION
**Pull request by chadwhitacre**
CI is failing on master. Why?

#1272

### CODE REVIEW

**Comment by chadwhitacre** [comment]
```
curl: (22) The requested URL returned error: 502 Bad Gateway
```

https://github.com/getsentry/self-hosted/runs/4929127714?check_suite_focus=true#step:4:1402

**Comment by chadwhitacre** [comment]
Restarting debug ...

**Comment by chadwhitacre** [comment]
Actually, does it fail locally? 🤔

**Comment by chadwhitacre** [comment]
Passes locally.

**Comment by chadwhitacre** [comment]
> curl: (22) The requested URL returned error: 502 Bad Gateway

https://github.com/getsentry/self-hosted/runs/5023770344?check_suite_focus=true#step:4:1378

That sounds like nginx is up but the app server is down. Why does it fail so late though? That's not the first time we hit the server, is it?

**Comment by BYK** [comment]
@chadwhitacre here is your error: https://github.com/getsentry/self-hosted/runs/5023770344?check_suite_focus=true#step:5:81

**Comment by aminvakil** [comment]
> @chadwhitacre here is your error: https://github.com/getsentry/self-hosted/runs/5023770344?check_suite_focus=true#step:5:81

So maybe giving #1267 a little love helps here too?
And why it has started to fail now? What has been changed?

**Comment by chadwhitacre** [comment]
Trying to repro locally ...

**Comment by chadwhitacre** [comment]
Based on https://github.com/getsentry/self-hosted/pull/1251#issuecomment-1015528084 I made an attempt to get our relay invocation to not spit out `credentials.json` properly, but it did. 🤔

```
$ curl -L https://github.com/docker/compose/releases/download/1.28.0/run.sh -o docker-compose-1.28.0
$
$
$
$ ./test-relay 
docker-compose version 1.28.0, build d02a7b1
docker-py version: 4.4.1
CPython version: 3.9.0
OpenSSL version: OpenSSL 1.1.1i  8 Dec 2020
Creating sentry-self-hosted_relay_run ... done
{"secret_key":"foo","public_key":"bar","id":"baz"}
$
$
$
$ cat test-relay
```

```sh
#!/usr/bin/env zsh

dc="./docker-compose-1.28.0"
$dc version
$dc run \
    --no-deps \
    --volume "$(pwd)/relay/config.yml:/tmp/config.yml" \
    relay --config /tmp credentials generate --stdout
```

**Comment by BYK** [comment]
@chadwhitacre let's add a `cat credentials.json` and an `ls` before that in inspect failure step to see what we generate on CI?

**Comment by chadwhitacre** [comment]
Ooooh, GitHub sez I won. 😜

Screen Shot 2022-02-01 at 11 54 48 AM

**Comment by chadwhitacre** [comment]
Hmm ... not seeing the cat? Guess we need that `ls` after all. :P

https://github.com/getsentry/self-hosted/runs/5024828441?check_suite_focus=true#step:4:520

File: install/relay-credentials.sh [hunk 1]
```diff
@@ -20,6 +21,10 @@ if [[ ! -f "$RELAY_CREDENTIALS_JSON" ]]; then
     > "$RELAY_CREDENTIALS_JSON"
```
Comment (BYK) [inline install/relay-credentials.sh:21]: Maybe we should add that `-T` here right after `$dcr`?

Comment (BYK) [reply install/relay-credentials.sh]: Because, read the giant comment I left there? 😁

Comment (chadwhitacre) [reply install/relay-credentials.sh]: I read it but I couldn't make sense of it.

> `relay generate credentials` tries to read the config and the credentials even with the `--stdout` and `--overwrite` flags and then errors out when the credentials file exists but not valid JSON. We hit this case as we redirect output to the same config folder, creating an empty credentials file before relay runs.

So no shell redirect means no credentials file created before relay runs means no error parsing JSON?

**Comment by chadwhitacre** [comment]
Okay! I did a `reset.sh` locally and now I am getting the same result as in CI:

> Volume sentry-data declared as external, but could not be found. Please create the volume manually using `docker volume create --name=sentry-data` and try again.

https://github.com/getsentry/self-hosted/runs/5027376491?check_suite_focus=true#step:4:19

**Comment by chadwhitacre** [comment]
Locally with 167082fe959aaf13ed1625460c66ae9aa185620b I am seeing "Read-only file system (os error 30)", [but in CI](https://github.com/getsentry/self-hosted/runs/5027800858?check_suite_focus=true#step:4:76) the relay run completes but does not store the file where I would expect.

```
$ ./test-relay 
docker-compose version 1.28.0, build d02a7b1
docker-py version: 4.4.1
CPython version: 3.9.0
OpenSSL version: OpenSSL 1.1.1i  8 Dec 2020
Removing network sentry-self-hosted_default
WARNING: Network sentry-self-hosted_default not found.
sentry-clickhouse
sentry-data
sentry-kafka
sentry-postgres
sentry-redis
sentry-symbolicator
sentry-zookeeper
total 8
-rw-r--r--  1 chadwhitacre  staff  317 Jun  9  2021 config.example.yml
-rw-r--r--  1 chadwhitacre  staff  324 Apr 15  2021 config.yml
Creating network "sentry-self-hosted_default" with the default driver
Creating volume "sentry-self-hosted_sentry-secrets" with default driver
Creating volume "sentry-self-hosted_sentry-smtp" with default driver
Creating volume "sentry-self-hosted_sentry-zookeeper-log" with default driver
Creating volume "sentry-self-hosted_sentry-kafka-log" with default driver
Creating volume "sentry-self-hosted_sentry-smtp-log" with default driver
Creating volume "sentry-self-hosted_sentry-clickhouse-log" with default driver
Creating sentry-self-hosted_relay_run ... done
 ERROR relay_log::utils > could not write config file (file /work/.relay/credentials.json)
  caused by: Read-only file system (os error 30)
ERROR: 1
total 8
-rw-r--r--  1 chadwhitacre  staff  317 Jun  9  2021 config.example.yml
-rw-r--r--  1 chadwhitacre  staff  324 Apr 15  2021 config.yml
cat: relay/credentials.json: No such file or directory
Creating sentry-self-hosted_relay_run ... done
 ERROR relay_log::utils > no stored credentials
ERROR: 1
$
```

**Comment by chadwhitacre** [comment]
Okay based on hints in https://github.com/getsentry/self-hosted/pull/1273#discussion_r796957245 I removed my local `config.yml` and now I see the same behavior as in CI. 👍

**Comment by chadwhitacre** [comment]
Okay! _Now_ I'm getting the read-only filesystem error I've been expecting:

https://github.com/getsentry/self-hosted/runs/5027923080?check_suite_focus=true#step:4:71

**Comment by chadwhitacre** [comment]
Discrepancy again between local and [CI](https://github.com/getsentry/self-hosted/runs/5027991448?check_suite_focus=true#step:4:88) with 6c2c1a952b98698b495e6c572ac8655a5a7ee88c. 🤔

```
$ ./test-relay 
docker-compose version 1.28.0, build d02a7b1
docker-py version: 4.4.1
CPython version: 3.9.0
OpenSSL version: OpenSSL 1.1.1i  8 Dec 2020
Removing sentry-self-hosted_relay_run_47c378141ef1 ... done
Removing sentry-self-hosted_relay_run_7606de44c280 ... done
Removing network sentry-self-hosted_default
sentry-clickhouse
sentry-data
sentry-kafka
sentry-postgres
sentry-redis
sentry-symbolicator
sentry-zookeeper
total 8
-rw-r--r--  1 chadwhitacre  staff  317 Jun  9  2021 config.example.yml
-rw-r--r--  1 chadwhitacre  staff  317 Feb  1 16:08 config.yml
Creating network "sentry-self-hosted_default" with the default driver
Creating sentry-self-hosted_relay_run ... 
Creating sentry-self-hosted_relay_run ... done
=========================================
total 8
-rw-r--r--  1 chadwhitacre  staff  317 Jun  9  2021 config.example.yml
-rw-r--r--  1 chadwhitacre  staff  317 Feb  1 16:08 config.yml
cat: relay/credentials.json: No such file or directory
Creating sentry-self-hosted_relay_run ... done
 ERROR relay_log::utils > no stored credentials
ERROR: 1
=========================================
total 12
-rw-r--r--  1 chadwhitacre  staff  317 Jun  9  2021 config.example.yml
-rw-r--r--  1 chadwhitacre  staff  317 Feb  1 16:08 config.yml
-rw-r--r--  1 chadwhitacre  staff  164 Feb  1 16:08 credentials.json
{"secret_key":"TdyNZalu4396gQQJWKgDWNNLbWDZgoKFm1KN4pqrjQ0","public_key":"I3_DyrtuIkY1J_rUMdAwV9qHNJFAPT1XmSVPr0ul-Rs","id":"4159e03e-4e7b-484a-b302-6f8f554ab3d4"}
Creating sentry-self-hosted_relay_run ... done
Credentials:
  relay id: 4159e03e-4e7b-484a-b302-6f8f554ab3d4
  public key: I3_DyrtuIkY1J_rUMdAwV9qHNJFAPT1XmSVPr0ul-Rs
$
```

**Comment by chadwhitacre** [comment]
Otoh this seems like it could be the root discrepancy between the CI failures on master and success in local. Varies also in Docker version? 🧐

**Comment by BYK** [comment]
Could it be file owner/permission issues?

**Comment by chadwhitacre** [comment]
> Could it be file owner/permission issues?

Maaaaaayyybbbeeeeeee????

https://github.com/getsentry/self-hosted/runs/5028581259?check_suite_focus=true#step:4:87 with 9ff6efb917c00329a73b1a83c2b0fff3e25cf58f indicates that the `credentials.json` file seems to look okay inside the container. 🧐


```
inside ----------------------------------
Creating sentry-self-hosted_relay_run ... 
Creating sentry-self-hosted_relay_run ... done
/work
total 12
drwxr-xr-x 3 relay 4096 Feb  1 21:52 ./
drwxr-xr-x 1 root  4096 Feb  1 21:52 ../
drwxr-xr-x 2  1001 4096 Feb  1 21:52 .relay/
total 20
drwxr-xr-x 2  1001 4096 Feb  1 21:52 ./
drwxr-xr-x 3 relay 4096 Feb  1 21:52 ../
-rw-r--r-- 1  1001  317 Feb  1 21:52 config.example.yml
-rw-r--r-- 1  1001  317 Feb  1 21:52 config.yml
-rw-r--r-- 1  1001  341 Feb  1 21:52 credentials.json
nightly: Pulling from getsentry/relay
Digest: sha256:c126666fd4f9a99388e276686eb5c81dd71a58c114c719b18a2f1d78653fd900
Status: Downloaded newer image for getsentry/relay:nightly
{"secret_key":"71_Jx4h2xEsyVy4gRlJuRKNR9vEq-5ocr1k0bgL4UAA","public_key":"QsyzZyXw0k-rWWES1RhXuC4n1OcTVFiPvjTy2EdhRoE","id":"20c1f5e6-f0fe-41c2-badc-484493048535"}
```

**Comment by BYK** [comment]
@chadwhitacre can you try `cat -v` instead? I suspect there is an invisible char at the beginning of the file, maybe the BOM header?

File: test-relay [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+cat relay/credentials.json
+$creds show
+
+echo 'moving ----------------------------------'
+mv credentials.json.tmp relay/credentials.json
+ls -FGl relay
+echo 'outside ---------------------------------'
+cat relay/credentials.json
+echo 'inside ----------------------------------'
+docker-compose run --no-deps relay bash -c 'whoami; pwd; ls -FGla; ls -FGla .relay; cat /work/.relay/credentials.json'
```
Comment (BYK) [inline test-relay:33]:
```suggestion
docker-compose run --no-deps relay bash -c 'whoami; pwd; ls -FGla; ls -FGla .relay; cat -v /work/.relay/credentials.json'
```

File: test-relay [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+cat relay/credentials.json
+$creds show
+
+echo 'moving ----------------------------------'
+mv credentials.json.tmp relay/credentials.json
+ls -FGl relay
+echo 'outside ---------------------------------'
+cat relay/credentials.json
+echo 'inside ----------------------------------'
+docker-compose run --no-deps relay bash -c 'whoami; pwd; ls -FGla; ls -FGla .relay; cat -v /work/.relay/credentials.json'
```
Comment (BYK) [inline test-relay:33]: Let's go further
```suggestion
docker-compose run --no-deps relay bash -c 'whoami; pwd; ls -FGla; ls -FGla .relay; cat -vE /work/.relay/credentials.json'
```

**Comment by chadwhitacre** [comment]
Worth trying but doesn't seem to be it. :-/

https://github.com/getsentry/self-hosted/runs/5028994154?check_suite_focus=true#step:4:105

Sure is a head-scratcher!

**Comment by BYK** [comment]
Okay found it, this is what is inside the file:
```
nightly: Pulling from getsentry/relay$
Digest: sha256:c126666fd4f9a99388e276686eb5c81dd71a58c114c719b18a2f1d78653fd900$
Status: Downloaded newer image for getsentry/relay:nightly$
{"secret_key":"WPqIzr_yCdueJNPgTdZlEisYeX80C3saSh1L2dEI9VI","public_key":"zAH4YGlW7Kmd-3oShwNWx_3SkuCo-6Odt2FEx5HLn6Y","id":"cf47dbb5-f70f-4ad1-9643-b9beb7ca38e5"}$
```

🤦🏻‍♂️

**Comment by chadwhitacre** [comment]
😭😭😭😭😭😭😭😭😭😭😭

**Comment by chadwhitacre** [comment]
Why do I not get bit by that locally?
Why does Docker Compose 2 work?

**Comment by BYK** [comment]
No idea but here's your fix:
```
--quiet-pull            Pull without printing progress information.
```

**Comment by chadwhitacre** [comment]
Dang. Green.

All the 🌮 @BYK 🙇 

🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮
🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮🌮

**Comment by BYK** [comment]
Okay so that only works with `docker compose` not `docker-compose`

**Comment by BYK** [comment]
Okay, I think the only reasonable fix here is to run `docker compose pull` upfront.

**Comment by chadwhitacre** [comment]
Okay that explains https://github.com/getsentry/self-hosted/runs/5029137872?check_suite_focus=true#step:4:163.

I can use the tail call in 5f67f7bb4c03f97bae0ce44a14e53b49e3018486.

**Comment by BYK** [comment]
Btw, `relay generate credentials` fails, right?

**Comment by chadwhitacre** [comment]
Does it? Link?

**Comment by BYK** [comment]
>Does it? Link?

Dunno, thought you tried that too :D

Btw setting `COMPOSE_INTERACTIVE_NO_CLI=1` may help with that output.

**Comment by chadwhitacre** [comment]
> Dunno, thought you tried that too :D

Ah, you mean without `--stdout`? Yes, I tried that somewhere back there and yes it failed you were right. ;)

**Comment by chadwhitacre** [comment]
Oh! Locally I already have the image cached.

**Comment by chadwhitacre** [comment]
Boom! 💥 

I removed the relay nightly image and I can repro locally:

```
relay/credentials.json ------------------
nightly: Pulling from getsentry/relay
Digest: sha256:c126666fd4f9a99388e276686eb5c81dd71a58c114c719b18a2f1d78653fd900
Status: Downloaded newer image for getsentry/relay:nightly
{"secret_key":"f66RDFJ-K-C6jHpsO7mGjuKXc_zXypdynQedNYzNmlU","public_key":"gyuqZ-p-TxVMj6xt0_aNDd8k8LBWsOmnR1RfISUH93I","id":"05775dda-6d9a-42ef-8ac8-c914ea9a3f33"}
=========================================
Creating sentry-self-hosted_relay_run ... done
error: could not parse json config file (file /work/.relay/credentials.json)
  caused by: expected ident at line 1 column 2
ERROR: 1
$
```

💃

**Comment by chadwhitacre** [comment]
> yes it failed you were right

But! We can work around the "redirect creates the empty file which confuses relay" issue by redirecting to a tmp file and moving that into place. So we're good there. 👍

Last thing I want to check is Docker Compose 2 behavior ...

**Comment by chadwhitacre** [comment]
Alright I'm outta time. I think with some fresh eyes tomorrow it shouldn't take too long to get a PR together.

Thanks for your help @BYK! 😁 💃 🌮 👏

**Comment by chadwhitacre** [comment]
Oh wow do the different CI runs use the same docker cache? 🤔

**Comment by chadwhitacre** [comment]
Because I'm surprised that 1.28.0 is passing 🤔

**Comment by chadwhitacre** [comment]
Looks like we need the `-T` for 1 and 2.2.3+. Okay! Gonna scrap this for parts. 💃

**Comment by aminvakil** [comment]
Wow! Glad it was sorted out at last!

**Pull request closed without merging**